Add generic base classes for WFS grids

geographika opened the pull request:
This pull request contains the following:

A view, view model, and controller for a generic WFS grid that can be reused in several applications. It contains attribute and spatial filtering, exports, paging and sorting. 

A new layer plug-in is also included to open a WFS grid associated with the layer.

Some minor edits have been made to the SpatialQueryButton to allow the layer to be set from the grid configuration. 

An example grid configuration with the model, store, viewmode, and grid view all in one file has been added `CpsiMapview.view.grid.ExampleGrid` to allow a demo in the cpsi-mapview project. 

A screencase showing the grid layer-plugin and the grid view with a linked spatial query tool is [available here](https://drive.google.com/file/d/1_euqALWDjRpLzv9Bz-kLrcJl5ojC-XH6/view).

chrismayer commented on app/view/grid/ExampleGrid.js:
```diff
@@ -0,0 +1,76 @@
+// create the grid view
+Ext.define('CpsiMapview.view.grid.ExampleGrid', {
+    extend: 'CpsiMapview.view.grid.Grid',
+    xtype: 'examplegrid',
```
A `cmv_` would be good here to keep the xtypes consistent

geographika replied:
Updated

chrismayer commented on app/view/grid/Grid.js:
```diff
@@ -0,0 +1,105 @@
+    viewConfig: {  //this config is passed to the view
+        loadMask: {
+            msg: 'Loading records' // TODO not sure why this isn't applied to the loadMask
+        }
```
Could be achieved with `loadingText: 'Loading records'`, see https://docs.sencha.com/extjs/6.2.0/classic/Ext.view.Table.html#cfg-loadingText

geographika replied:
I've tried various properties such as `loadingText `, `loadMessage` in and out of the `viewConfig` with no luck. 
I can set it with `view.setLoading('Loading Records...');` in `onWfsStoreBeforeLoad` with no problems though.

chrismayer commented on app/view/menuitem/LayerGrid.js:
```diff
@@ -0,0 +1,80 @@
+        var title = me.layer.get('name');
+
+        // we can't keep a reference to the window in this class
+        // as a new Ext.menu.Item i created each time the menu is
```
typo ` ... is created ...`

geographika replied:
Fixed

chrismayer commented on app/controller/grid/Grid.js:
```diff
@@ -0,0 +1,328 @@
+        // TODO check for feature type when zooming
+        var extent = feature.getGeometry().getExtent();
+        // as this is a point then buffer it by 100m
+        extent = ol.extent.buffer(extent, 100);
```
Making the buffer width configurable in the grid would be a good idea, IMHO.

geographika replied:
Added as config option

chrismayer commented on app/controller/grid/Grid.js:
```diff
@@ -0,0 +1,328 @@
+                    text: 'Zoom to Feature',
+                    scope: me,
+                    handler: function () {
+                        var record = grid.getSelection()[0];
```
This should be the `record` variable provided in the `onItemContextMenu` listener function.

geographika replied:
The record always seems to be available when `onItemContextMenu` is called but is then undefined in the handler. I can set me.record in the `onItemContextMenu` function and access this in the handler, but this doesn't seem the cleanest approach. I'm unsure why the variable isn't available in the anonymous function.

geographika replied:
Updated

chrismayer commented on app/controller/grid/Grid.js:
```diff
@@ -0,0 +1,328 @@
+        grid.saveDocumentAs({
+            type: 'xlsx',
+            title: 'Map Export',
+            fileName: 'myExport.xlsx'
```
Would be good to have these properties as config in the grid view class

geographika replied:
Added

chrismayer commented on app/controller/grid/Grid.js:
```diff
@@ -0,0 +1,328 @@
+     * Fire a onRowDblClick to be handled by the parent application
+     */
+    onRowDblClick: function (sender, record) {
+        this.fireEvent('onRowDblClick', sender, record);
```
Is it really necessary to forward the `rowdblclick` event of the grid?

geographika replied:
In a child application I used this approach, but I guess I can listen to the event directly from a view without needing to forward anything. Will remove from here.

geographika commented:
@chrismayer - many thanks for the review! I've addressed the issues. There are likely a few more options that will become configurable over time as this base grid is used in more applications. 
The `loadingText` remains a mystery to me, but currently a minor display issue. 
I've fixed the spatial filter paging bug - thanks for catching that. 
Will merge tomorrow.

chrismayer commented:
Thanks for your ongoing effort @geographika, very nice.

Regarding to the `loadingText`: Agree, if it does not work as described in the API-Docs it is maybe a bug in ExtJS. So if `view.setLoading('Loading Records...')` works then it is fine of course.
